### PR TITLE
Better exceptions (particularly for syntax errors)

### PR DESCRIPTION
**Pull request by ry**

### CODE REVIEW

**Comment by kitsonk** [comment]
LGTM

File: libdeno/libdeno_test.cc [hunk 1]
```diff
@@ -240,14 +243,16 @@ TEST(LibDenoTest, EncodeErrorBug) {
...
-               "{\"message\":\"ReferenceError: a is not defined\","
-               "\"frames\":[{\"line\":1,\"column\":1,"
-               "\"functionName\":\"\",\"scriptName\":\"<unknown>\","
-               "\"isEval\":true,"
-               "\"isConstructor\":false,\"isWasm\":false},{\"line\":1,"
-               "\"column\":1,\"functionName\":\"\",\"scriptName\":\"a.js\","
-               "\"isEval\":false,\"isConstructor\":false,\"isWasm\":false}]}");
+  EXPECT_STREQ(
+      deno_last_exception(d),
+      "{\"message\":\"ReferenceError: a is not "
```
Comment (piscisaureus) [inline libdeno/libdeno_test.cc]: This is such an unreadable mess...

File: libdeno/libdeno_test.cc [hunk 2]
```diff
@@ -228,10 +229,12 @@ TEST(LibDenoTest, LastException) {
   EXPECT_EQ(deno_last_exception(d), nullptr);
   EXPECT_FALSE(deno_execute(d, nullptr, "a.js", "\n\nthrow Error('boo');\n\n"));
   EXPECT_STREQ(deno_last_exception(d),
-               "{\"message\":\"Error: boo\","
-               "\"frames\":[{\"line\":3,\"column\":7,"
-               "\"functionName\":\"\",\"scriptName\":\"a.js\","
-               "\"isEval\":false,"
+               "{\"message\":\"Error: boo\",\"sourceLine\":\"throw "
+               "Error('boo');\",\"scriptResourceName\":\"a.js\",\"lineNumber\":"
```
Comment (piscisaureus) [inline libdeno/libdeno_test.cc]: Also unreadable mess...

File: libdeno/libdeno_test.cc [hunk 3]
```diff
@@ -172,13 +172,14 @@ TEST(LibDenoTest, SnapshotBug) {
...
   Deno* d = deno_new(deno_config{0, snapshot, empty, nullptr, nullptr});
   EXPECT_FALSE(deno_execute(d, nullptr, "a.js", "GlobalErrorHandling()"));
-  // We only check that it starts with this string, so we don't have to check
-  // the second frame, which contains line numbers in libdeno_test.js and may
-  // change over time.
   std::string expected =
       "{\"message\":\"ReferenceError: notdefined is not defined\","
-      "\"frames\":[{\"line\":3,\"column\":2,\"functionName\":\"\","
-      "\"scriptName\":\"helloworld.js\",\"isEval\":true,"
+      "\"sourceLine\":\" "
```
Comment (piscisaureus) [inline libdeno/libdeno_test.cc]: Also unreadable mess...

File: libdeno/binding.cc [hunk 4]
```diff
@@ -83,6 +83,66 @@ std::string EncodeExceptionAsJSON(v8::Local<v8::Context> context,
   // auto exception_str = message->Get();
   CHECK(json_obj->Set(context, v8_str("message"), exception_str).FromJust());
 
+  auto maybe_source_line = message->GetSourceLine(context);
```
Comment (piscisaureus) [inline libdeno/binding.cc]: This whole section is unfortunately quite un-DRY. Maybe we can refactor it later somehow (an X macro maybe?)

Comment (ry) [reply libdeno/binding.cc]: I tried it at first with a macro but it seemed more complexity than it was worth.

**Comment by kevinkassimo** [comment]
Found a small problem:
Given a file `bad_import.ts`:
```ts
import "non-existent.txt";
```
The new error message given by `deno bad_import.ts` turns out to be:
```
gen/bundle/main.js:4339:17
          return new DenoError(kind, base.error());

NotFound: Cannot resolve module "non-existent.txt" from "/Users/kevinqian/Desktop/Programming/Deno/test/bad_import.ts"
    at DenoError (deno/js/errors.ts:19:5)
    at maybeError (deno/js/errors.ts:38:12)
    at maybeThrowError (deno/js/errors.ts:26:15)
    at sendSync (deno/js/dispatch.ts:67:5)
    at codeFetch (deno/js/os.ts:46:19)
    at resolveModule (deno/js/compiler.ts:357:38)
    at moduleNames.map.name (deno/js/compiler.ts:519:31)
    at resolveModuleNames (deno/js/compiler.ts:511:24)
    at compilerHost.resolveModuleNames (deno/third_party/node_modules/typescript/lib/typescript.js:32750:71)
    at resolveModuleNamesWorker (deno/third_party/node_modules/typescript/lib/typescript.js:20688:69)
```
This `gen/bundle/main.js:4339:17
          return new DenoError(kind, base.error());` would seem confusing to users...

**Pull request merged**